File: networking/src/main/java/work/fking/pangya/networking/protocol/ProtocolDecoder.java
package work.fking.pangya.networking.protocol;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufUtil;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ByteToMessageDecoder;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import work.fking.pangya.networking.crypt.PangCrypt;

import java.util.List;

public class ProtocolDecoder extends ByteToMessageDecoder {

    private static final Logger LOGGER = LogManager.getLogger(InboundPacketDispatcher.class);

    private final Protocol protocol;
    private final int cryptKey;

    private ProtocolDecoder(Protocol protocol, int cryptKey) {
        this.protocol = protocol;
        this.cryptKey = cryptKey;
    }

    public static ProtocolDecoder create(Protocol protocol, int cryptKey) {
        return new ProtocolDecoder(protocol, cryptKey);
    }

    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out) {
        PangCrypt.decrypt(buffer, cryptKey);

        int packetId = buffer.readShortLE();
        LOGGER.trace("Incoming packetId=0x{}", Integer.toHexString(packetId));

        try {
            InboundPacket inboundPacket = protocol.createInboundPacket(packetId, buffer);

            if (inboundPacket == null) {
                LOGGER.warn("Unknown packetId=0x{}", Integer.toHexString(packetId));
                LOGGER.warn("\n{}", ByteBufUtil.prettyHexDump(buffer));
                buffer.clear();
                ctx.disconnect();
                return;
            }
            out.add(inboundPacket);
        } catch (Throwable throwable) {
            throw new ProtocolException(throwable);
        }
    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/net/ProtocolDecoder.kt
package work.fking.pangya.game.net

import io.netty.buffer.ByteBuf
import io.netty.buffer.ByteBufUtil
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.ByteToMessageDecoder
import org.apache.logging.log4j.LogManager
import work.fking.pangya.networking.crypt.PangCrypt

private val LOGGER = LogManager.getLogger(ProtocolDecoder::class.java)

class ProtocolDecoder(
    private val protocol: ClientProtocol,
    private val cryptKey: Int
) : ByteToMessageDecoder() {

    override fun decode(ctx: ChannelHandlerContext, buffer: ByteBuf, out: MutableList<Any>) {
        PangCrypt.decrypt(buffer, cryptKey)
        val packetId = buffer.readShortLE().toInt()
        val packetType = protocol.forId(packetId)
        LOGGER.debug("Incoming packet {} (0x{})", packetType, Integer.toHexString(packetId))

        if (packetType == null) {
            LOGGER.warn("Unknown packetId=0x{}", Integer.toHexString(packetId))
            LOGGER.warn("\n{}", ByteBufUtil.prettyHexDump(buffer))
            ctx.disconnect()
            buffer.clear()
            return
        }
        out.add(ClientPacket(packetType, buffer.readRetainedSlice(buffer.readableBytes())))
    }
}
